src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -525,28 +525,37 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
 
                 asyncState = family == AddressFamily.Unspecified ? (object)ipAddress : new KeyValuePair<IPAddress, AddressFamily>(ipAddress, family);
             }
-            else if (NameResolutionPal.SupportsGetAddrInfoAsync)
+            else
             {
+                if (NameResolutionPal.SupportsGetAddrInfoAsync)
+                {
 #pragma warning disable CS0162 // Unreachable code detected -- SupportsGetAddrInfoAsync is a constant on *nix.
 
-                // If the OS supports it and 'hostName' is not an IP Address, resolve the name asynchronously
-                // instead of calling the synchronous version in the ThreadPool.
+                    // If the OS supports it and 'hostName' is not an IP Address, resolve the name asynchronously
+                    // instead of calling the synchronous version in the ThreadPool.
 
-                ValidateHostName(hostName);
+                    ValidateHostName(hostName);
 
-                if (NameResolutionTelemetry.Log.IsEnabled())
-                {
-                    return justAddresses
-                        ? (Task)GetAddrInfoWithTelemetryAsync<IPAddress[]>(hostName, justAddresses, family, cancellationToken)
-                        : (Task)GetAddrInfoWithTelemetryAsync<IPHostEntry>(hostName, justAddresses, family, cancellationToken);
-                }
-                else
-                {
-                    return NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses, family, cancellationToken);
+                    Task t;
+                    if (NameResolutionTelemetry.Log.IsEnabled())
+                    {
+                        t = justAddresses
+                            ? (Task)GetAddrInfoWithTelemetryAsync<IPAddress[]>(hostName, justAddresses, family, cancellationToken)
+                            : (Task)GetAddrInfoWithTelemetryAsync<IPHostEntry>(hostName, justAddresses, family, cancellationToken);
+                    }
+                    else
+                    {
+                        t = NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses, family, cancellationToken);
+                    }
+
+
+                    if (!t.IsFaulted || !(t.Exception?.InnerException is InvalidOperationException))
```

Comment:
This check for `InvalidOperationException` is intended to work with specifically the exception added in `NameResolutionPal.Windows.cs` for `errorCode == SocketError.TryAgain`, or generally with any `InvalidOperationException`s that are possibly thrown from the depth of `GetAddrInfoAsync`?

Reply:
In my first round I simply raised the original error and look for particular socket error here. However there are other cases when this can happen:

> WSATRY_AGAIN 11002 | Nonauthoritative host not found.This is usually a temporary error during host name resolution and means that the local server did not receive a response from an authoritative server. A retry at some time later may be successful.

So I wanted to pin the retry only for the case where we suspect it is caused by the impersonification. If we for example try again in case when nam server is unreachable, we would only incur extra delay.

To avoid this, I map the synchronous WSATRY_AGAIN to exception I think is very unlikely to happen otherwise and do the retry only in that case. Errors coming from the GetAddrInfoExW will be mapped to SocketException.